crates/bevy_ecs/src/entity/entity_set.rs

```diff
@@ -0,0 +1,443 @@
+use alloc::{
+    collections::{btree_map, btree_set},
+    rc::Rc,
+    sync::Arc,
+};
+
+use core::{
+    array,
+    fmt::{Debug, Formatter},
+    iter::{self, FusedIterator},
+    option, result,
+};
+
+use super::Entity;
+
+/// A trait for entity borrows.
+///
+/// This trait can be thought of as `Borrow<Entity>`, but yielding `Entity` directly.
+pub trait EntityBorrow {
```

Comment:
I was imagining that a few calls to `Iterator::copied()` could prevent the deeply-nested types and save a whole pile of `impl` blocks.  Not a big deal either way!  

Hmm, given that you aren't returning a reference anymore, maybe `AsEntity` from the linked issue is a better name than `EntityBorrow`.  

Oh, and that reminds me!  Do you want to implement `TrustedEntityBorrow` for `MainEntity` and `RenderEntity`, since those are simple newtypes over `Entity`?

Reply:
> Oh, and that reminds me! Do you want to implement TrustedEntityBorrow for MainEntity and RenderEntity, since those are simple newtypes over Entity?

Yes! Anything that represents an `Entity` and can delegate its comparison trait to `Entity`.
There's a handful more of `Entity` newtypes in bevy itself as far as I'm aware.
`EntityBorrow` could even be made derivable.